In [1]:
import numpy as np
from scipy import stats
import scipy
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest

In [2]:
# slide #67 paired t-test
#df1, df2를 직접 입력
df1 = pd.DataFrame({'before':[720,589,780,648,720,589,780,648,780,648]})
df2 = pd.DataFrame({'after':[810,670,790,712,810,670,790,712,790,712]})
#paired t-test
t_result = stats.ttest_rel(df1,df2)
#paired t-test값 p값 저장
t,p = t_result.statistic.round(3),t_result.pvalue.round(3)
#paired t-test값 p값 출력
print("Paired t-test")
print("t검정통계량: {}".format(t))
print("p-value: {}".format(p))

Paired t-test
t검정통계량: [-5.324]
p-value: [0.]


### 결론: 유의수준 5%에서 검정결과 P값이 0.000이므로 학원 수강 전후의 점수는 차이가 있다고 할 수 있다.

In [3]:
# slide #68 paired t-test 실습
df1 = pd.DataFrame({'before':[720,589,780,648,720,589,780,648,780,648]})
df2 = pd.DataFrame({'after':[710,580,787,712,750,600,782,670,790,680]})
t_result = stats.ttest_rel(df1,df2)
t,p = t_result.statistic.round(3),t_result.pvalue.round(4)
print("Paired t-test")
print("t검정통계량: {}".format(t))
print("p-value: {}".format(p))

Paired t-test
t검정통계량: [-2.266]
p-value: [0.0497]


### 결론: 유의수준 5%에서 검정한 결과 p-value가 0.0497이므로 귀무가설을 기각할 수 없다. 그래서 학원 수강 후의 성적에 차이가 있다고 할 수 없다.

In [4]:
# slide #69 ANOVA 연습
df1 = pd.DataFrame({'A':[892,623,721,678,723,790,720,670,690,771]})
df2 = pd.DataFrame({'B':[721,821,910,678,723,790,711,790,745,891]})
df3 = pd.DataFrame({'C':[621,915,888,721,894,834,841,912,845,889]})
#ANOVA 실행
f_result = stats.f_oneway(df1,df2,df3)
#ANOVA의 f값 p값 저장
f,p = f_result.statistic.round(3),f_result.pvalue.round(3)
#ANOVA의 f값 p값 출력
print("One-way ANOVA")
print("F검정통계량: {}".format(f))
print("p-value: {}".format(p))

One-way ANOVA
F검정통계량: [4.263]
p-value: [0.025]


In [5]:
# slide #70 1 Proportion test 연습
from statsmodels.stats.proportion import proportions_ztest
#변수 선언
count = 15
nobs = 100
value = .1
#Proportion test 실행
stat,pval = proportions_ztest(count,nobs,value)
#Proportion test 결과 출력
print("1 Proportion test")
print("p검정통계량: {0:0.3f}".format(stat))
print("p-value: {0:0.3f}".format(pval))

1 Proportion test
p검정통계량: 1.400
p-value: 0.161


### P-value가 0.161로 귀무가설을 기각할 수 없다. 그러므로 품질개선 결과로 기존보다 전체 고객 중 만족도 차이가 있다고 할 수 없다.

In [6]:
from statsmodels.stats.proportion import proportions_ztest
#변수 선언
count = 10
nobs = 100
value = .2
#Proportion test 실행
stat,pval = proportions_ztest(count,nobs,value)
#Proportion test 결과 출력
print("1 Proportion test")
print("p검정통계량: {0:0.3f}".format(stat))
print("p-value: {0:0.3f}".format(pval))

1 Proportion test
p검정통계량: -3.333
p-value: 0.001


### P-value가 0.001로 귀무가설을 기각한다. 그러므로 품질개선 결과로 기존보다 전체 고객 중 만족도 차이가 있다.

In [7]:
# slide #72 2 Proportion test 연습
#변수 선언
count = np.array([4,1])
nobs = np.array([1000,1200])
#Proportion test 실행
stat,pval = proportions_ztest(count,nobs)
#Proportion test 결과 출력
print("2 Proportion test")
print("p검정통계량: {0:0.3f}".format(stat))
print("p-value: {0:0.3f}".format(pval))

2 Proportion test
p검정통계량: 1.553
p-value: 0.120


### H0 : p1(불량률을 의미) = p2, H1 : p1 != p2 p-value가 0.120으로 유의수준보다 크므로 귀무가설을 기각할 수 없다. 
### 그러므로 두 공정의 불량률은 차이가 있다고 할 수 없다.

In [8]:
# slide #73 2 Proportion test 연습
#변수 선언
count = np.array([14,5])
nobs = np.array([1200,1200])
#Proportion test 실행
stat,pval = proportions_ztest(count,nobs)
#Proportion test 결과 출력
print("2 Proportion test")
print("p검정통계량: {0:0.3f}".format(stat))
print("p-value: {0:0.3f}".format(pval))

2 Proportion test
p검정통계량: 2.073
p-value: 0.038


### H0 : p1(불량률을 의미) = p2, H1 : p1 != p2 p-value가 0.038으로 유의수준보다 작으므므로 귀무가설을 기각한다. 
### 그러므로 두 공정의 불량률은 차이가 있다.

In [9]:
df = pd.DataFrame({'HSG':[270,260,236,234],'SS':[228,285,225,262],'SPA':[277,284,231,208]})
#chi-square test 실행
chi,pval,dof,expected = stats.chi2_contingency(df.T)
#chi-square test 결과 출력
print("chi-square test")
print("chisq:{0:0.3f}".format(chi))
print("p:{0:0.3f}".format(pval))
print('degree of freedom: {}'.format(dof))
print('expected value:\n{}'.format(expected.round(3)))

chi-square test
chisq:13.366
p:0.038
degree of freedom: 6
expected value:
[[258.333 276.333 230.667 234.667]
 [258.333 276.333 230.667 234.667]
 [258.333 276.333 230.667 234.667]]


### 결론: 근무조별로 제품을 생산하는 부하 차이가 있다고 할 수 있다.

In [10]:
df = pd.DataFrame({'AS':[18,8,4,4,3,3],'가격':[1,2,1,1,1,25],'성능':[8,14,3,2,3,8],'확장성':[7,5,4,3,1,10],'디자인':[10,5,9,2,1,2],'안정성':[9,9,5,7,1,1],'기능성':[10,4,4,3,1,7]})
#chi-square test 실행
chi,pval,dof,expected = stats.chi2_contingency(df.T)
#chi-square test 결과 출력
print("chi-square test")
print("chisq:{0:0.3f}".format(chi))
print("p:{0:0.3f}".format(pval))
print('degree of freedom: {}'.format(dof))
print('expected value:\n{}'.format(expected.round(3)))

chi-square test
chisq:98.274
p:0.000
degree of freedom: 30
expected value:
[[11.004  8.21   5.24   3.843  1.921  9.782]
 [ 8.528  6.362  4.061  2.978  1.489  7.581]
 [10.454  7.799  4.978  3.651  1.825  9.293]
 [ 8.253  6.157  3.93   2.882  1.441  7.336]
 [ 7.978  5.952  3.799  2.786  1.393  7.092]
 [ 8.803  6.568  4.192  3.074  1.537  7.825]
 [ 7.978  5.952  3.799  2.786  1.393  7.092]]


In [11]:
df1 = [1,2,3,4,4,5,6,6,7,8]
df2 = [23,29,49,64,74,87,96,97,109,119]
#Correlation Analysis 실행
corr,pval = stats.pearsonr(df1, df2)
print("Correlation Analysis")
print("corr: {0:0.3f}".format(corr))
print("p-value: {0:0.3f}".format(pval))

Correlation Analysis
corr: 0.989
p-value: 0.000


### H0: 부품수리시간과 부품수 간에 유의미한(선형) 관계가 없다. H1: 부품수리시간과 부품수 간에 유의미한(선형) 관계가 있다.
### p가 0.05보다 작아서 귀무가설 기각. 유의미한 관계가 있고 r이 0.989로 강한 상관관계를 가진다고 할 수 있다.

In [12]:
df1 = [37.6,38.6,37.2,36.4,38.6,39,37.2,36.1,35.9,37.1,36.9,37.5,36.3,38.1,39,36.9,36.8,37.6,33,33.5]
df2 = [14,15,14,16,17,14,17,16,15,16,14,16,18,13,15,17,14,16,20,21]
#Correlation Analysis 실행
corr,pval = stats.pearsonr(df1, df2)
print("Correlation Analysis")
print("corr: {0:0.5f}".format(corr))
print("p-value: {0:0.5f}".format(pval))

Correlation Analysis
corr: -0.72311
p-value: 0.00032


In [13]:
MINUTES = [1,2,3,4,4,5,6,6,7,8]
UNITS = [23,29,49,64,74,87,96,97,109,119]

#Simple Linear Regression 실행
MINUTES = sm.add_constant(MINUTES)
model = sm.OLS(UNITS, MINUTES)
result = model.fit()
#출력
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.979
Model:                            OLS   Adj. R-squared:                  0.976
Method:                 Least Squares   F-statistic:                     365.3
Date:                Fri, 25 Jan 2019   Prob (F-statistic):           5.82e-08
Time:                        15:12:35   Log-Likelihood:                -29.401
No. Observations:                  10   AIC:                             62.80
Df Residuals:                       8   BIC:                             63.41
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.1712      3.886      1.845      0.1

/home/pirl/anaconda3/lib/python3.7/site-packages/scipy/stats/stats.py:1394: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  "anyway, n=%i" % int(n))


In [14]:
STRENGTH = [37.6,38.6,37.2,36.4,38.6,39,37.2,36.1,35.9,37.1,36.9,37.5,36.3,38.1,39,36.9,36.8,37.6,33,33.5]
BAKING_TIME = [0.798922,0.84902,0.816163,0.758266,0.815894,0.807477,0.809068,0.7664,0.742243,0.751118,0.740629,0.751495,0.738863,0.827428,0.818172,0.740939,0.757128,0.804695,0.695121,0.735377]
#Simple Linear Regression 실행
STRENGTH = sm.add_constant(STRENGTH)
model = sm.OLS(BAKING_TIME, STRENGTH)
result = model.fit()
#출력
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.646
Model:                            OLS   Adj. R-squared:                  0.627
Method:                 Least Squares   F-statistic:                     32.90
Date:                Fri, 25 Jan 2019   Prob (F-statistic):           1.95e-05
Time:                        15:12:35   Log-Likelihood:                 46.530
No. Observations:                  20   AIC:                            -89.06
Df Residuals:                      18   BIC:                            -87.07
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0023      0.135      0.017      0.9

In [1]:
import pandas as pd
import numpy as np
from numpy.random import randn
#샘플 데이터 통일을 위해 seed 고정: 1234
np.random.seed(1234)
df = pd.DataFrame(randn(10,2), columns = ['A','B'])

In [2]:
df

A         B
0  0.471435 -1.190976
1  1.432707 -0.312652
2 -0.720589  0.887163
3  0.859588 -0.636524
4  0.015696 -2.242685
5  1.150036  0.991946
6  0.953324 -2.021255
7 -0.334077  0.002118
8  0.405453  0.289092
9  1.321158 -1.546906

In [17]:
df['A'] #숫자가 순서대로 출력됨 Series형태
df[['A']] #표 안에 출력 DataFrame 형태
df[0:5] #0~4행 자료가 나옴.

A         B
0  0.471435 -1.190976
1  1.432707 -0.312652
2 -0.720589  0.887163
3  0.859588 -0.636524
4  0.015696 -2.242685

In [18]:
df.loc[0:7,['A']] # A열 0~7행 선택
df.loc[:,['A']] # A열 전체
df.loc[3:8] #3~8행 긁어옴
df.loc[3:3] # 3행만

A         B
3  0.859588 -0.636524

In [19]:
# loc와 차이점은 index로 뽑아오기 때문
df.iloc[0:7,[0]] # A열 0~6행 선택
df.iloc[:,[0]] # A열 전체
df.iloc[3:8] #3~7행 긁어옴
df.iloc[3:4] #3행만

A         B
3  0.859588 -0.636524

In [3]:
df.query('A>0.5') #조건을 이용한 데이터 선택./ A열에 0.5보다 큰 데이터들만 뽑아줌.
df.query('A>B')
#query와 동일한 slice 표현
df[df.A>df.B]

A         B
0  0.471435 -1.190976
1  1.432707 -0.312652
3  0.859588 -0.636524
4  0.015696 -2.242685
5  1.150036  0.991946
6  0.953324 -2.021255
8  0.405453  0.289092
9  1.321158 -1.546906

In [21]:
np.random.seed(1234)
df = pd.DataFrame(randn(10,4), columns = ['A','B','C','D'])

In [22]:
df.iloc[0:7,:]

A         B         C         D
0  0.471435 -1.190976  1.432707 -0.312652
1 -0.720589  0.887163  0.859588 -0.636524
2  0.015696 -2.242685  1.150036  0.991946
3  0.953324 -2.021255 -0.334077  0.002118
4  0.405453  0.289092  1.321158 -1.546906
5 -0.202646 -0.655969  0.193421  0.553439
6  1.318152 -0.469305  0.675554 -1.817027

In [12]:
df1 = pd.DataFrame({'A':['A0','A1','A2','A3'], 'B':['B0','B1','B2','B3'], 'C':['C0','C1','C2','C3'],'D':['D0','D1','D2','D3']},index=[0,1,2,3])
df2 = pd.DataFrame({'A':['A4','A5','A6','A7'], 'B':['B4','B5','B6','B7'], 'C':['C4','C5','C6','C7'],'D':['D4','D5','D6','D7']},index=[4,5,6,7])
df3 = pd.DataFrame({'A':['A8','A9','A10','A11'], 'B':['B8','B9','B10','B11'], 'C':['C8','C9','C10','C11'],'D':['D8','D9','D10','D11']},index=[8,9,10,11])
df4 = pd.DataFrame({'A':['A0','A1','A2','A3','A4','A5','A6','A7','A8','A9','A10','A11'], 'B':['B0','B1','B2','B3','B4','B5','B6','B7','B8','B9','B10','B11'], 'C':['C0','C1','C2','C3','C4','C5','C6','C7','C8','C9','C10','C11'], 'D':['D0','D1','D2','D3','D4','D5','D6','D7','D8','D9','D10','D11']})
df4

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

In [24]:
result = pd.concat([df1,df2,df3]) # default axis = 0 세로결합
result

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

In [14]:
result = pd.concat([df1,df2,df3],axis = 1) #가로결합
result
result = pd.concat([df1,df2,df3],axis = 1, ignore_index = True) # 열 필드가 ABC에서 0123형태로 바뀜
result

0    1    2    3    4    5    6    7    8    9    10   11
0    A0   B0   C0   D0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1    A1   B1   C1   D1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2    A2   B2   C2   D2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
3    A3   B3   C3   D3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
4   NaN  NaN  NaN  NaN   A4   B4   C4   D4  NaN  NaN  NaN  NaN
5   NaN  NaN  NaN  NaN   A5   B5   C5   D5  NaN  NaN  NaN  NaN
6   NaN  NaN  NaN  NaN   A6   B6   C6   D6  NaN  NaN  NaN  NaN
7   NaN  NaN  NaN  NaN   A7   B7   C7   D7  NaN  NaN  NaN  NaN
8   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A8   B8   C8   D8
9   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A9   B9   C9   D9
10  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A10  B10  C10  D10
11  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A11  B11  C11  D11

In [5]:
left = pd.DataFrame({'key1':['K0','K1','K2','K3'], 'key2':['K0','K1','K0','K1'], 'A':['A0','A1','A2','A3'], 'B':['B0','B1','B2','B3']})
right = pd.DataFrame({'key1':['K0','K1','K2','K3'], 'key2':['K0','K0','K0','K0'], 'C':['C0','C1','C2','C3'],'D':['D0','D1','D2','D3']})
print(left)
print(right)
result = pd.merge(left,right,on=['key1','key2'])
result

  key1 key2   A   B
0   K0   K0  A0  B0
1   K1   K1  A1  B1
2   K2   K0  A2  B2
3   K3   K1  A3  B3
  key1 key2   C   D
0   K0   K0  C0  D0
1   K1   K0  C1  D1
2   K2   K0  C2  D2
3   K3   K0  C3  D3


key1 key2   A   B   C   D
0   K0   K0  A0  B0  C0  D0
1   K2   K0  A2  B2  C2  D2

In [6]:
result = pd.merge(left,right,on=['key1','key2'],how='left')
result

key1 key2   A   B    C    D
0   K0   K0  A0  B0   C0   D0
1   K1   K1  A1  B1  NaN  NaN
2   K2   K0  A2  B2   C2   D2
3   K3   K1  A3  B3  NaN  NaN

In [7]:
result = pd.merge(left,right,on=['key1','key2'],how='right')
result

key1 key2    A    B   C   D
0   K0   K0   A0   B0  C0  D0
1   K2   K0   A2   B2  C2  D2
2   K1   K0  NaN  NaN  C1  D1
3   K3   K0  NaN  NaN  C3  D3

In [8]:
result = pd.merge(left,right,on=['key1','key2'],how ='left')
result

key1 key2   A   B    C    D
0   K0   K0  A0  B0   C0   D0
1   K1   K1  A1  B1  NaN  NaN
2   K2   K0  A2  B2   C2   D2
3   K3   K1  A3  B3  NaN  NaN

In [9]:
result = pd.merge(left,right,on=['key1','key2'],how='left',indicator=True)
result

key1 key2   A   B    C    D     _merge
0   K0   K0  A0  B0   C0   D0       both
1   K1   K1  A1  B1  NaN  NaN  left_only
2   K2   K0  A2  B2   C2   D2       both
3   K3   K1  A3  B3  NaN  NaN  left_only